### PR TITLE
Implement a simple SchemaDescription visitor

### PR DESCRIPTION
**Pull request by daddykotex**
I'm not sure this is what was expected, mostly because I figure the solution is much more verbose than the current solution. On the other side, this visitor will be available for other usages if we need it.

### CODE REVIEW

File: modules/core/src/smithy4s/internals/SchemaDescription.scala [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+ *  limitations under the License.
+ */
+
+package smithy4s
+package internals
+
+import smithy4s.schema.SchemaVisitor
+import smithy4s.schema.{Primitive, EnumValue, SchemaField, SchemaAlt, Alt}
+
+trait SchemaDescription[F] {
```
Comment (Baccata) [inline modules/core/src/smithy4s/internals/SchemaDescription.scala:23]: You can make this a type alias in the internals package : 

```
type SchemaDescription[A] = String
```

File: modules/core/src/smithy4s/internals/SchemaDescription.scala [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+    SchemaDescription.of(s"Enumeration")
+  
+  override def struct[S](shapeId: ShapeId, hints: Hints, fields: Vector[SchemaField[S, _]], make: IndexedSeq[Any] => S): SchemaDescription[S] =
+    SchemaDescription.of(s"Structure")
+  
+  override def union[U](shapeId: ShapeId, hints: Hints, alternatives: Vector[SchemaAlt[U, _]], dispatch: U => Alt.SchemaAndValue[U, _]): SchemaDescription[U] =
+    SchemaDescription.of(s"Union")
+  
+  override def biject[A, B](schema: Schema[A], to: A => B, from: B => A): SchemaDescription[B] =
+    SchemaDescription.of(s"Bijection")
```
Comment (Baccata) [inline modules/core/src/smithy4s/internals/SchemaDescription.scala:75]: This should return the description of the `schema` parameter

File: modules/core/src/smithy4s/internals/SchemaDescription.scala [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+  override def surject[A, B](schema: Schema[A], to: Refinement[A,B], from: B => A): SchemaDescription[B] =
```
Comment (Baccata) [inline modules/core/src/smithy4s/internals/SchemaDescription.scala:76]: ditto

File: modules/core/src/smithy4s/internals/SchemaDescription.scala [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+    val valueSc = apply(value)
+    SchemaDescription.of(s"Map[${keySc.description}, ${valueSc.description}]")
+  }
+  override def enumeration[E](shapeId: ShapeId, hints: Hints, values: List[EnumValue[E]], total: E => EnumValue[E]): SchemaDescription[E] =
+    SchemaDescription.of(s"Enumeration")
+  
+  override def struct[S](shapeId: ShapeId, hints: Hints, fields: Vector[SchemaField[S, _]], make: IndexedSeq[Any] => S): SchemaDescription[S] =
+    SchemaDescription.of(s"Structure")
+  
+  override def union[U](shapeId: ShapeId, hints: Hints, alternatives: Vector[SchemaAlt[U, _]], dispatch: U => Alt.SchemaAndValue[U, _]): SchemaDescription[U] =
```
Comment (Baccata) [inline modules/core/src/smithy4s/internals/SchemaDescription.scala:71]: This should give something really similar to the `structure` case, with `union` instead

File: modules/core/src/smithy4s/internals/SchemaDescription.scala [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+  }
+  override def map[K, V](shapeId: ShapeId, hints: Hints, key: Schema[K], value: Schema[V]): SchemaDescription[Map[K,V]] = {
+    val keySc = apply(key)
+    val valueSc = apply(value)
+    SchemaDescription.of(s"Map[${keySc.description}, ${valueSc.description}]")
+  }
+  override def enumeration[E](shapeId: ShapeId, hints: Hints, values: List[EnumValue[E]], total: E => EnumValue[E]): SchemaDescription[E] =
+    SchemaDescription.of(s"Enumeration")
+  
+  override def struct[S](shapeId: ShapeId, hints: Hints, fields: Vector[SchemaField[S, _]], make: IndexedSeq[Any] => S): SchemaDescription[S] =
```
[the remark this answers is not in the document]
Comment (daddykotex) [reply modules/core/src/smithy4s/internals/SchemaDescription.scala:68]: I was not sure if we wanted to unfold the more complex types

[the remark this answers is not in the document]
Comment (Baccata) [reply modules/core/src/smithy4s/internals/SchemaDescription.scala:68]: I'm happy for us not too, but in this case you have to be consistent and not unfold the lists/maps

[the remark this answers is not in the document]
Comment (daddykotex) [reply modules/core/src/smithy4s/internals/SchemaDescription.scala:68]: I'll move to the simple version for now, but will explore the more complex version (unfolding the shape and carrying some state for the `lazy` one)

**Comment by daddykotex** [comment]
I'll mark this as draft, I'm not able to implement `lazy` with the `Set[ShapeId] => (Set[ShapeId], String)` approach

File: modules/core/src/smithy4s/internals/SchemaDescription.scala [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+  override def map[K, V](shapeId: ShapeId, hints: Hints, key: Schema[K], value: Schema[V]): SchemaDescriptionDetailed[Map[K,V]] = {
+    apply(key).flatMapResult { kDesc =>
+      apply(value).mapResult { vDesc =>
+        s"Map[$kDesc, $vDesc]"
+      }
+    }
+  }
+  override def enumeration[E](shapeId: ShapeId, hints: Hints, values: List[EnumValue[E]], total: E => EnumValue[E]): SchemaDescriptionDetailed[E] = {
+    val vDesc = values.map(e => e.stringValue).mkString(", ")
+    SchemaDescriptionDetailed.of(shapeId, s"Enumeration{ $vDesc }")
```
Comment (Baccata) [inline modules/core/src/smithy4s/internals/SchemaDescription.scala:109]: inconsistent whitespace between `Enumeration` and the bracket. For instance, in `Structure`, there's a whitespace between the structure name and the bracket. 

I'm happy for the whitespace to be removed over there.

File: modules/core/src/smithy4s/internals/package.scala [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+package smithy4s
+
+package object internals {
+  type SchemaDescription[A] = String
+  // type SchemaDescriptionDetailed[A] = Set[ShapeId] => (Set[ShapeId], String)
```
Comment (Baccata) [inline modules/core/src/smithy4s/internals/package.scala:21]: remove this ?

**Review by Baccata — APPROVED** [review]
A few nitpicks that I'd like addressed before merging, lgtm otherwise

**Comment by daddykotex** [comment]
Remarking as draft, I just merged into my other branch and I'm not satisfied. Everything works fine, but the `Set[ShapeId]` is leaking out and it should not

File: modules/core/src/smithy4s/schema/SchemaVisitor.scala [hunk 8]
```diff
@@ -42,4 +42,27 @@ trait SchemaVisitor[F[_]] extends (Schema ~> F) {
     case SurjectionSchema(schema, to, from) => surject(schema, to, from)
     case LazySchema(make) => lazily(make)
   }
+
+  def mapK[G[_]](f: F ~> G): SchemaVisitor[G] = {
```
Comment (Baccata) [inline modules/core/src/smithy4s/schema/SchemaVisitor.scala:46]: To remove. `Schema ~> F` already has `andThen` which achieves exactly the same thing. 

https://github.com/disneystreaming/smithy4s/blob/main/modules/core/src/smithy4s/PolyFunction.scala#L28-L31

Comment (Baccata) [reply modules/core/src/smithy4s/schema/SchemaVisitor.scala]: Addressed in [73ec53e](https://github.com/disneystreaming/smithy4s/pull/295/commits/73ec53ebac13c50817bb5f799584856b840199d5)

**Pull request merged**